src/cpp/ext/otel/otel_plugin.cc

```diff
@@ -536,6 +627,209 @@ void OpenTelemetryPlugin::RecordHistogram(
+  if (absl::holds_alternative<Disabled>(instrument_data.instrument)) {
+    // This instrument is disabled.
+    return;
+  }
+  GPR_ASSERT(absl::holds_alternative<std::unique_ptr<ObservableState<int64_t>>>(
+      instrument_data.instrument));
+  const auto& descriptor =
+      grpc_core::GlobalInstrumentsRegistry::GetInstrumentDescriptor(handle);
+  GPR_ASSERT(descriptor.label_keys.size() == label_values.size());
+  GPR_ASSERT(descriptor.optional_label_keys.size() == optional_values.size());
+  auto& observable_state = absl::get<std::unique_ptr<ObservableState<int64_t>>>(
+      instrument_data.instrument);
+  {
+    grpc_core::MutexLock l{&observable_state->mu};
+    observable_state->value = value;
+    observable_state->label_values = std::vector<absl::string_view>(
+        label_values.begin(), label_values.end());
+    observable_state->optional_label_values = std::vector<absl::string_view>(
+        optional_values.begin(), optional_values.end());
+  }
+  if (!std::exchange(observable_state->callback_registered, true)) {
+    observable_state->instrument->AddCallback(&ObservableCallback<int64_t>,
+                                              observable_state.get());
+  }
+}
+void OpenTelemetryPlugin::SetGauge(
+    grpc_core::GlobalInstrumentsRegistry::GlobalDoubleGaugeHandle handle,
+    double value, absl::Span<const absl::string_view> label_values,
+    absl::Span<const absl::string_view> optional_values) {
+  const auto& instrument_data = instruments_data_.at(handle.index);
+  if (absl::holds_alternative<Disabled>(instrument_data.instrument)) {
+    // This instrument is disabled.
+    return;
+  }
+  GPR_ASSERT(absl::holds_alternative<std::unique_ptr<ObservableState<double>>>(
+      instrument_data.instrument));
+  const auto& descriptor =
+      grpc_core::GlobalInstrumentsRegistry::GetInstrumentDescriptor(handle);
+  GPR_ASSERT(descriptor.label_keys.size() == label_values.size());
+  GPR_ASSERT(descriptor.optional_label_keys.size() == optional_values.size());
+  auto& observable_state = absl::get<std::unique_ptr<ObservableState<double>>>(
+      instrument_data.instrument);
+  {
+    grpc_core::MutexLock l{&observable_state->mu};
+    observable_state->value = value;
+    observable_state->label_values = std::vector<absl::string_view>(
+        label_values.begin(), label_values.end());
+    observable_state->optional_label_values = std::vector<absl::string_view>(
+        optional_values.begin(), optional_values.end());
+  }
+  if (!std::exchange(observable_state->callback_registered, true)) {
+    observable_state->instrument->AddCallback(&ObservableCallback<double>,
+                                              observable_state.get());
+  }
+}
+
+void OpenTelemetryPlugin::AddCallback(
+    grpc_core::RegisteredMetricCallback* callback) {
+  {
+    grpc_core::MutexLock l{&mu_};
+    callback_states_.emplace(callback, RegisteredMetricCallbackState{});
+  }
+  for (const auto& handle : callback->metrics()) {
```

Comment:
You mean some other thread call `RemoveCallback()` passing the same `RegisteredMetricCallback*` as the argument? I don't think we want to allow that usage. If that thread's `RemoveCallback()` is executed first, it should just crash. It would not help if we lock the whole functions in that situation.